src/MLMCManager.cpp

```diff
@@ -0,0 +1,450 @@
+{
+    double out = 0.0;
+    for (const auto& v : mean_)
+    {
+        out += v;
+    }
+    return out;
+}
+
+/**
+   This ugly implementation is to use Vector::Swap which does not
+   exist for BlockVector, so we wrap Vector in temporary BlockVector
+   objects that share the data with Vector.
+*/
+mfem::BlockVector InterpolateToFine(const Upscale& upscale, int level,
+                                    const mfem::BlockVector& in)
+{
+    mfem::Vector vec1, vec2;
+    vec1 = in;
+    for (int k = level; k > 0; k--)
+    {
+        MFEM_ASSERT(vec1.Size() == upscale.GetMatrix(k).GetBlockOffsets().Last(),
+                    "Sizes do not work!");
+        mfem::BlockVector block_vec1(vec1.GetData(),
+                                     upscale.GetMatrix(k).GetBlockOffsets());
+        vec2.SetSize(upscale.GetMatrix(k - 1).GetBlockOffsets().Last());
+        mfem::BlockVector block_vec2(vec2.GetData(),
+                                     upscale.GetMatrix(k - 1).GetBlockOffsets());
+        /// Interpolate from k to the finer k-1
+        upscale.Interpolate(k, block_vec1, block_vec2);
+        vec2.Swap(vec1);
+    }
+    mfem::BlockVector out(upscale.GetMatrix(0).GetBlockOffsets());
+    MFEM_ASSERT(out.Size() == vec1.Size(), "Sizes do not work!");
+    // ((mfem::Vector) out) = vec1; // doesn't work for some reason (valgrind doesn't like it)
+    for (int i = 0; i < out.Size(); ++i)
+    {
+        out[i] = vec1[i];
+    }
+    return out;
+}
+
+/// restrict from finest level (level 0) to given level
+mfem::BlockVector RestrictToLevel(const Upscale& upscale,
+                                  int level, const mfem::BlockVector& in)
+{
+    mfem::Vector vec1, vec2;
+    vec1 = in;
+    for (int k = 0; k < level; ++k)
+    {
+        MFEM_ASSERT(vec1.Size() == upscale.GetMatrix(k).GetBlockOffsets().Last(),
+                    "Sizes do not work!");
+        mfem::BlockVector block_vec1(vec1.GetData(),
+                                     upscale.GetMatrix(k).GetBlockOffsets());
+        vec2.SetSize(upscale.GetMatrix(k + 1).GetBlockOffsets().Last());
+        vec2 = 0.0; // ????
+        mfem::BlockVector block_vec2(vec2.GetData(),
+                                     upscale.GetMatrix(k + 1).GetBlockOffsets());
+        upscale.Restrict(k + 1, block_vec1, block_vec2);
+        vec2.Swap(vec1);
+    }
+    mfem::BlockVector out(upscale.GetMatrix(level).GetBlockOffsets());
+    MFEM_ASSERT(out.Size() == vec1.Size(), "Sizes do not work!");
+    // ((mfem::Vector) out) = vec1;
+    for (int i = 0; i < out.Size(); ++i)
+    {
+        out[i] = vec1[i];
+    }
+    return out;
+}
+
+void MLMCManager::CoarseSample(bool verbose)
+{
+    const int level = num_levels_ - 1;
+    const int pressure_block = 1;
+
+    sampler_.NewSample();
+    auto coarse_coefficient = sampler_.GetCoefficient(level);
+    fvupscale_.RescaleCoefficient(level, coarse_coefficient);
+    mfem::BlockVector rhs_coarse = RestrictToLevel(fvupscale_, level, rhs_fine_);
```

Comment:
Whenever a new coefficient is sampled, `rhs_fine_` and `rhs_coarse` are not changed. So this restriction step can be performed once, and the rhs for all levels are stored for repeated use (both here and `CorrectionSample`.